### ISSUE TITLE
AGIC HTTPS listener does not work as expected

### ISSUE THREAD

**Issue by bgvladedivac** [Azure/application-gateway-kubernetes-ingress#817]
**Describe the bug**

It's not exactly a bug, it could be just me not having a clear understanding, but do not hurry up to close it please :)

In our K8s cluster we have ArgoCD and it only accepts HTTPs connections. The SSL Certificate that the pod exposes is issued by a CA. 

In the cluster there's a pre-created TLS secret that encapsulates the TLS certificate + private key. 

My understanding is that those are the prerequisites for me to have HTTPS listener on APPGW that would pass all traffic to the ArgoCD pod on the backend. It's a little bit challenging to find information in your annotation list how exactly to provision HTTPS listener, so any improvement there would be great. I might do it on my own, if manage to understand the whole process.

My Ingress:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: azure-ingress
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/backend-protocol: https
    appgw.ingress.kubernetes.io/ssl-redirect: "true"
spec:
  tls:
    - secretName: secret-tls
  rules:
  - http:
      paths:
      - backend:
          serviceName: argocd-server
          servicePort: 443
        path: /*
```
Once I try to access the AG I receive:

```
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>Microsoft-Azure-Application-Gateway/v2</center>
</body>
</html>
```

**To Reproduce**
1. Install AGIC.
2. Deploy TLS secret with certificate + private key.
3. Install ArgoCD 
4. Create the above referenced Ingress manifest.

**Ingress Controller details**
```
PriorityClassName:  <none>
Start Time:         Wed, 22 Apr 2020 16:43:02 +0300
Labels:             app=ingress-azure
                    pod-template-hash=6b65cd5f9b
                    release=azure-tests-fluentd
Annotations:        prometheus.io/port: 8123
                    prometheus.io/scrape: true
Status:             Running
Controlled By:      ReplicaSet...
Containers:
  ingress-azure:
    Container ID:   docker://301ae97a364e7f85f3fbededf2b10fd2898773ff97b73f2d2669ffb344b987a1
    Image:          mcr.microsoft.com/azure-application-gateway/kubernetes-ingress:1.0.0
    Image ID:       docker-pullable://mcr.microsoft.com/azure-application-gateway/kubernetes-ingress@sha256:c295f99ae66443c5a392fd894620fcd1fc313b9efdec96d13f166fefb29780a9
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Wed, 22 Apr 2020 16:43:10 +0300
    Ready:          True
    Restart Count:  0
    Liveness:       http-get http://:8123/health/alive delay=15s timeout=1s period=20s #success=1 #failure=3
    Readiness:      http-get http://:8123/health/ready delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment Variables from:
      azure-tests-fluentd-cm-ingress-azure  ConfigMap  Optional: false
    Mounts:
      /etc/Azure/Networking-AppGW/auth from networking-appgw-k8s-azure-service-principal-mount (ro)
      /etc/appgw/azure.json from azure (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from azure-tests-fluentd-sa-ingress-azure-token-ccjxg (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  azure:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/kubernetes/azure.json
    HostPathType:  File
  networking-appgw-k8s-azure-service-principal-mount:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  networking-appgw-k8s-azure-service-principal
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

 ```
kubectl logs <ingress controller>. 
I0423 14:37:33.041525       1 mutate_app_gateway.go:154] BEGIN AppGateway deployment
I0423 14:37:55.111608       1 mutate_app_gateway.go:182] Applied App Gateway config in 22.070057016s
I0423 14:37:55.111631       1 mutate_app_gateway.go:198] cache: Updated with latest applied config.
I0423 14:37:55.111946       1 mutate_app_gateway.go:203] END AppGateway deployment
I0423 14:39:02.562081       1 mutate_app_gateway.go:154] BEGIN AppGateway deployment
I0423 14:40:14.890419       1 mutate_app_gateway.go:182] Applied App Gateway config in 1m12.328313851s
I0423 14:40:14.890441       1 mutate_app_gateway.go:198] cache: Updated with latest applied config.
I0423 14:40:14.891036       1 mutate_app_gateway.go:203] END AppGateway deployment
I0423 14:40:15.985674       1 mutate_app_gateway.go:154] BEGIN AppGateway deployment
I0423 14:40:38.212205       1 mutate_app_gateway.go:182] Applied App Gateway config in 22.22650301s
I0423 14:40:38.212227       1 mutate_app_gateway.go:198] cache: Updated with latest applied config.
I0423 14:40:38.212827       1 mutate_app_gateway.go:203] END AppGateway deployment


```
All the workloads are deployed into the same namespace. Some of the logs are skipped, due to security, but in none of them there was something disturbing.

**Comment by 3quanfeng** [comment]
Hi @bgvladedivac, If I understand right, you are trying to set up[ E2E SSL encryption,](https://docs.microsoft.com/en-us/azure/application-gateway/ssl-overview) did you set up your trusted root certificate in your HttoSettings?

**Comment by bgvladedivac** [comment]
Hi @3quanfeng thanks for the reply.

The merged certificate ( merging root + intermediate ) is provided in the kubernetes secret. However, I am getting this error. Assuming it happens, because it's not just the intermediate certificate, if I upload just intermediate there's no error. It makes sence, since based on kubectl docs they support just intermediate certificates.

`E0427 08:24:05.488894       1 secretstore.go:111] unable to export using openssl, error=[exit status 1], stderr=[No certificate matches private key`

kubectl output:

```
The public/private key pair must exist before hand. The public key certificate must be .PEM encoded and match the given private key.

Examples:
  # Create a new TLS secret named tls-secret with the given key pair:
  kubectl create secret tls tls-secret --cert=path/to/tls.cert --key=path/to/tls.key
```
If I can't provide my certificate that way, how am I supposed to create HTTPS listener out of the box using the Ingress Controller ? 

In the HttpSettings the only option for a certificate that I see is 

> Use well known CA certificate

that is enabled.

Something that managed to fix was that the CN of the backend certificate was not matching the host header entered in the health probe configuration.

Still getting this 502 status back. I am heading to this conclusion, that there are a lot of manual tweaks that I must do pre/post do achieve something straightforward in my opinion.

**Comment by bgvladedivac** [comment]
Ok, I saw that you can't upload a trusted root certificate using the Portal.

Also I saw that you have updated some of the annotations, but they are updated only in your github README, while they are missing on the first reliable source https://azure.github.io/application-gateway-kubernetes-ingress/annotations/.

I have found the precioues annotation:
`appgw.ingress.kubernetes.io/appgw-trusted-root-certificate: "name-of-my-root-cert1"`

However, during the creation of the root certificate I end up with:

`$ az.cmd network application-gateway root-cert create --verbose --cert-file ca_bundle.cer  --gateway-name some-gateway --name some-root-cert-name --resource-group some-resoure-grup
ERROR: Either Data or KeyVaultSecretId must be specified for Certificate '/subscriptions/...
INFO: command ran in 2.006 seconds`

This leads to #https://github.com/MicrosoftDocs/azure-docs/issues/33157 

In a nutshell, my KeyVault covers the soft-delete enabled, retention period = 90 days + all networks could access it.

When you were doing the tests for this annotation, how did u manage to import the root certificate by the cli ?

**Comment by bgvladedivac** [comment]
Hi again :)

@akshaysngupta I saw that you are self-assigned so just wanted to share my progress.

I am following your docs under [https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/annotations.md#appgw-trusted-root-certificate](url) and to be more precise the section of 

> Configure your root certificate to Application Gateway

`
mv test.crt test.cer
`
`
az network application-gateway root-cert create --cert-file test.cer  --gateway-name $appgwName  --name name-of-my-root-cert1 --resource-group $resgp`

The example actually refers to a self signed certificate with a missing root certificate.

`
openssl verify -show_chain test.crt
...
error 18 at 0 depth lookup: self signed certificate
error test.crt: verification failed
`

The current status is that on my import attempts for a root certificate I always end up with an Invalid Data for the certificate. 

My assumptions:

The root certificate must not be full chained and PFX encoded, since that would require a password and I do not see any password field. 

In your example you are referring to a *.cer file which covers PEM. 

I tried with pure PEM and base64 encoded PEM, again same output.

**Comment by bgvladedivac** [comment]
Hi one more and sorry I am being a little bit too persistent in my self - answers :)

The only way to import the certificate is by using this guide and Microsoft Tools https://docs.microsoft.com/en-us/azure/application-gateway/certificates-for-backend-authentication.

Now I am having this seems like a final brick statement = >  the 'root certificate of the server certificate used by the backend, does not match the trusted root certificate added to the App GW'.

Does that mean that the root certificate also must be binded to the ArgoCD pod ? In my case ArgoCD is my backend, and it just exposes a secret with tls.crt + tls.key. If I try to incorporate the root + tls certificate into 1 key in the secret I am getting back a private key error and ArgoCD initializes with a self signed certificate.

Could you just confirm if in End to End encryption scenario the trusted root certificate also must be found in the backend ? And if that's the case, why's that ?

**Comment by akshaysngupta** [comment]
@bgvladedivac In order for SSL handshake to complete successfully, backend Pod must send the entire except the root certificate as root certificate is already configured on the client (in this case Application Gateway).

Also, you should make sure that Application Gateway is using the right host name when talking to the backend. This can be configured through [backend-hostname](https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/annotations.md#backend-hostname) annotation.
This host name is used for both Server name indication/host header when communicating to the Pod. This should exactly match the subject name of the certificate. This is similar to how a browser matches the certificate subject name to the address you type in the location bar during validation of the certificate.

1. `AppGW --- < backend-hostname >: 443 ---> Pod`
Initiate SSL handshake to the Pod using `< backend-hostname >` as SNI
1. `AppGW <---(Certificate Chain. root is optional)  (subject name)---Pod`
AppGW checks whether cert subject name == `< backend-hostname >` and check the certificate is trusted by the root certificate starting from the intermediate.
1. `AppGW ---HTTP Request with Host: < backend-hostname > header--->Pod`

**Comment by niteenkole** [comment]
Hi Akshay,

Any help appreciated we in situation where we cannot proceed.

We in same situation as above and not sure what we doing wrong.

1.We have entrust signed certificate and applying to our backend.
2.Our backend has full chain of certificates.
3.we applied root certificate to AGIC and use appgw.ingress.kubernetes.io/appgw-trusted-root-certificate: "nv-xxxxxx-root-cert1"
4.We using right hostname appgw.ingress.kubernetes.io/backend-hostname: "nv-xxxxxx.ca"

**Comment by niteenkole** [comment]
No worries I have it working.

Few things,when we export root through windows it has some special characters used dos2unix
have to use full chain of certificates for ingress and backend

**Comment by barlass** [comment]
Did anyone use LetsEncrypt ?

**Comment by S0meth1ng85** [comment]
Hi, I have the same problem. Trying to configure AppGW AGIC with end to end SSL. Our application is running on port 5556 and uses TLS https://pod-ip:5556

Ingress is configured in terraform:

`resource "kubernetes_ingress" "ingress_datapower" {
  count = var.deploy_datapower_ingress ? 1 : 0

  metadata {
    name      = lookup(var.services[count.index], "name")
    namespace = "default"
    annotations = {
      "kubernetes.io/ingress.class" = "azure/application-gateway"
      "appgw.ingress.kubernetes.io" = "https"
      "appgw.ingress.kubernetes.io/backend-hostname" = "datapower.gfieast.com"
    }
  }

  spec {

    tls {
      secret_name = kubernetes_secret.datapower_tls_secret.metadata.0.name
    }

    backend {
      service_name = lookup(var.services[count.index], "name")
      service_port = lookup(var.services[count.index], "target_port")
    }
    rule {
      http {
        path {
          path = "/"
          backend {
            service_name = lookup(var.services[count.index], "name")
            service_port = lookup(var.services[count.index], "target_port")
          }
        }
      }
    }
  }

  depends_on = [kubernetes_service.datapower_service]
  
}`

SSL certificate is installed on our application so we can access it via name https://wildcard.cert.name:5556

In Kubernetes secret we have SSL chain for the same wild certificate and private key as installed on application.
Don`t know why but AppGW listener has port 443 and not 5556In the HTTP Settings Backend protocol is HTTP and not HTTPS.
When I am trying to change it manually I can uplaod root CA cert. BUt it does not help...

Do MS has manual  how to setup end to end SSL with AppGW AGIC?
